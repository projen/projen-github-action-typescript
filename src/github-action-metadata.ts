import {
  Input,
  Output,
  JavaScriptRuns,
  DockerRuns,
  CompositeRuns,
  Branding,
} from './model/actions-metadata-model';

export enum Type {
  STRING = 'string',
  NUMBER = 'number',
  STRING_LIST = 'string[]',
  NUMBER_LIST = 'number[]',
  JSON = 'Record<string, string>',
}

export interface ProjenInput extends Input {
  /**
   * PROJEN ONLY: Specify the type of input. This will be used to
   * generate sample code, and is not part of GitHub Input Syntax.
   *
   * @default Type.STRING
   */
  readonly type?: Type;
}

/**
 * A Github Action Metadata file definition.
 */
export interface GitHubActionMetadata {
  /**
   * The name of your action. GitHub displays the `name` in the Actions tab to
   * help visually identify actions in each job.
   *
   * @default this.name
   */
  readonly name?: string;

  /**
   * The name of the action's author.
   *
   * @default - none
   */
  readonly author?: string;

  /**
   * A short description of the action.
   *
   * @default - a basic description is provided for you
   */
  readonly description?: string;

  /**
   * Input parameters allow you to specify data that the action expects to use
   * during runtime. GitHub stores input parameters as environment variables.
   * Input ids with uppercase letters are converted to lowercase during runtime.
   *
   * @default {}
   */
  readonly inputs?: Record<string, ProjenInput>;

  /**
   * Output parameters allow you to declare data that an action sets. This is
   * useful for actions that run later in a workflow, as they can use the output
   * data set in previously run actions.
   *
   * If you don't declare an output in your action metadata file, you can still
   * set outputs and use them in a workflow.
   * @see https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter
   *
   * @default {}
   */
  readonly outputs?: Record<string, Output>;

  /**
   * Specifies whether this is a JavaScript action, a composite action, or a Docker
   * container action and how the action is executed.
   */
  readonly runs: JavaScriptRuns | CompositeRuns | DockerRuns;

  /**
   * You can use a color and a Feather icon to create a badge to personalize and
   * distinguish your action. Badges are shown next to your action name in
   * GitHub Marketplace.
   *
   * @see https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#branding
   *
   * @default - no branding
   */
  readonly branding?: Branding;
}
