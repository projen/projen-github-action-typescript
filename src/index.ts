export * from './project';
export * from './github-action-metadata';
export * from './model/actions-metadata-model';