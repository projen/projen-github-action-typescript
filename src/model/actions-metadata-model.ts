import { Step } from 'projen/lib/github/workflows-model';

/**
 * A Github Action Metadata file definition.
 */
export interface ActionMetadata {
  /**
   * The name of your action. GitHub displays the `name` in the Actions tab to
   * help visually identify actions in each job.
   */
  readonly name: string;

  /**
   * The name of the action's author.
   *
   * @default - none
   */
  readonly author?: string;

  /**
   * A short description of the action.
   */
  readonly description: string;

  /**
   * Input parameters allow you to specify data that the action expects to use
   * during runtime. GitHub stores input parameters as environment variables.
   * Input ids with uppercase letters are converted to lowercase during runtime.
   *
   * @default {}
   */
  readonly inputs?: Record<string, Input>;

  /**
   * Output parameters allow you to declare data that an action sets. This is
   * useful for actions that run later in a workflow, as they can use the output
   * data set in previously run actions.
   *
   * If you don't declare an output in your action metadata file, you can still
   * set outputs and use them in a workflow.
   * @see https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter
   *
   * @default {}
   */
  readonly outputs?: Record<string, Output>;

  /**
   * Specifies whether this is a JavaScript action, a composite action, or a Docker
   * container action and how the action is executed.
   */
  readonly runs: JavaScriptRuns | CompositeRuns | DockerRuns;

  /**
   * You can use a color and a Feather icon to create a badge to personalize and
   * distinguish your action. Badges are shown next to your action name in
   * GitHub Marketplace.
   *
   * @see https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#branding
   *
   * @default - no branding
   */
  readonly branding?: Branding;
}

/**
 * Define an input to a GitHub Action
 */
export interface Input {
  /**
   * A description of the input parameter
   */
  readonly description: string;

  /**
   * Indicate whether the action requires the input parameter.
   *
   * @default false
   */
  readonly required?: boolean;

  /**
   * The default value when the input parameter is not specified
   * in the workflow file. This is required if `required=true`.
   */
  readonly default?: string;

  /**
   * If this parameter is set, it will be logged as a warning message
   * if the input parameter is used.
   *
   * @default - none
   */
  readonly deprecationMessage?: string;
}

/**
 * Define an output in a GitHub Action
 */
export interface Output {
  /**
   * The description of the output parameter.
   */
  readonly description: string;

  /**
   * The value that the output parameter will be mapped to.
   * This property is required if you are using composite actions,
   * and must be omitted otherwise.
   */
  readonly value?: string;
}

/**
 * The GitHub Action runtime or composite/docker identifier.
 */
export enum RunsUsing {
  /**
   * "node12"
   */
  NODE_12 = 'node12',

  /**
   * "node16"
   */
  NODE_16 = 'node16',

  /**
   * "composite"
   */
  COMPOSITE = 'composite',

  /**
   * "docker"
   */
  DOCKER = 'docker',
}

/**
 * Common interface for the Runs property.
 */
interface Runs {
  /**
   * The runtime used to execute the specified code.
   *
   * For JavaScript actions:
   * - use `RunsUsing.NODE_12` for Node.js v12.
   * - use `RunsUsing.NODE_16` for Node.js v16.
   *
   * For composite actions:
   * - use `RunsUsing.COMPOSITE`
   *
   * For docker container actions:
   * - use `RunsUsing.DOCKER`
   */
  readonly using: RunsUsing;
}

/**
 * Runs property for a JavaScript Action.
 */
export interface JavaScriptRuns extends Runs {
  /**
   * The file that contains your action code. The runtime specified in the
   * `using` property executes this file.
   */
  readonly main: string;

  /**
   * Allows you to run a script at the start of a job, before the `main`
   * action begins.
   *
   * @default - no pre steps
   */
  readonly pre?: string;

  /**
   * Allows you to define conditions for the `pre` action execution. The
   * `pre` action will only run if the conditions in `preIf` are met.
   *
   * For example, you can include the condition `runner.os == 'linux'` and
   * `pre` will only run on Linux-based runners.
   *
   * @defualt - no conditions
   */
  readonly preIf?: string;

  /**
   * Allows you to run a script at the end of a job, once the `main` action
   * has completed.
   *
   * @default - no post steps
   */
  readonly post?: string;

  /**
   * Allows you to define conditions for the `post` action execution. The
   * `post` action will only run if the conditions in `postIf` are met.
   *
   * For example, you can include the condition `runner.os == 'linux'` and
   * `post` will only run on Linux-based runners.
   *
   * @default - no conditions
   */
  readonly postIf?: string;
}

/**
 * Runs property for a Docker container action.
 */
export interface DockerRuns extends Runs {
  /**
   * Allows you to run a script before the `entrypoint` action begins.
   *
   * @default - no pre script
   */
  readonly preEntrypoint?: string;

  /**
   * Overrides the Docker `ENTRYPOINT` in the Dockerfile, or sets it if one
   * was not already specified.
   *
   * @default - uses `ENTRYPOINT` instruction in Dockerfile.
   */
  readonly entrypoint?: string;

  /**
   * Allows you to run a cleanup script once `entrypoint` action has completed.
   *
   * @default - no post script
   */
  readonly postEntrypoint?: string;

  /**
   * The Docker image to use as the container to run the action.
   */
  readonly image: string;

  /**
   * Specifies a key/value map of environment variables to set in the container
   * environment.
   */
  readonly env?: Record<string, string>;
}

/**
 * Runs property for a composite action.
 */
export interface CompositeRuns extends Runs {
  /**
   * The steps that you plan to run in a composite action.
   */
  readonly steps: ActionStep[];
}

/**
 * Definition for branding the GitHub Action.
 */
export interface Branding {
  /**
   * The background color of the badge. Can be one of:
   * - `white`
   * - `yellow`
   * - `blue`
   * - `green`
   * - `orange`
   * - `red`
   * - `purple`
   * - `gray-dark`
   */
  readonly color: string;

  /**
   * The name of the v4.28.0 Feather icon to use. Brand icons are omitted.
   * See link for additional omitted icons as well as an exhaustive list of
   * currently supported icons:
   *
   * @see https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#brandingicon
   */
  readonly icon: string;
}

/**
 * Definition for a step used in a composite action.
 */
export interface ActionStep extends Step {
  /**
   * Which shell to use for running the step.
   *
   * @example "bash"
   */
  readonly shell?: string;
}
