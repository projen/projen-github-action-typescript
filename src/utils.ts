export function upperCase(input: string): string {
  const wordList = splitWords();
  console.log(wordList);
  for (let i = 0; i < wordList.length; i++) {
    wordList[i] = wordList[i][0].toLocaleUpperCase() + wordList[i].slice(1);
  }
  return wordList.join('');

  function splitWords() {
    var regex = /[A-Z\xC0-\xD6\xD8-\xDE]?[a-z\xDF-\xF6\xF8-\xFF]+|[A-Z\xC0-\xD6\xD8-\xDE]+(?![a-z\xDF-\xF6\xF8-\xFF])|\d+/g;
    return input.match(regex) ?? [];
  }
}