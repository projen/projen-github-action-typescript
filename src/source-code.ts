import { Component } from 'projen';
import { GitHubActionMetadata } from './github-action-metadata';
import { Type } from './model/actions-metadata-model';
import { GitHubActionTypeScriptProject } from './project';

interface PropType {
  readonly name: string;
  readonly type: Type;
  readonly required: boolean;
  readonly default?: string;
  readonly description: string;
}

export class GitHubActionSourceCode extends Component {
  private includeListFunction = false;
  private includeNumberFunction = false;
  private properties: Record<string, PropType> = {};

  /**
   * Code for index.ts
   */
  public indexCode: string[];

  /**
   * Code for action-options.ts
   */
  public actionOptionsCode: string[];

  constructor(project: GitHubActionTypeScriptProject) {
    super(project);

    this.getProperties(project.actionMetadata);

    this.indexCode = [
      'import * as core from \'@actions/core\';',
      'import { ActionManager } from \'./action\';',
      '',
      'async function run() {',
      ...this.getInputs(indent(1)),
      '  const action = new ActionManager({',
      ...this.renderProperties(indent(2)),
      '  });',
      '',
      '  await action.main();',
      '}',
      ...renderListFunction(this.includeListFunction),
      ...renderNumberFunction(this.includeNumberFunction),
      '',
      'run().catch(error => {',
      '  core.setFailed(error.message);',
      '});',
    ];

    this.actionOptionsCode = [
      'export interface ActionManagerOptions {',
      ...this.renderOptions(indent(1)),
      '}',
    ];

  }

  private getProperties(actionMetadata: GitHubActionMetadata) {
    for (const [key, input] of Object.entries(actionMetadata.inputs ?? {})) {
      this.properties[key] = {
        name: sanitize(key),
        type: input.type ?? Type.STRING,
        required: input.required ?? false,
        default: input.default,
        description: input.description,
      };
    }
  }

  private renderProperties(ind: string): string[] {
    const renderedProps = [];
    for (const { name } of Object.values(this.properties)) {
      renderedProps.push(`${ind}${name},`);
    }
    return renderedProps;
  }

  private getInputs(ind: string): string[] {
    const result = [];
    for (const [key, { name, type }] of Object.entries(this.properties)) {
      result.push(`${ind}const ${name}: ${type} = ${this.getTypedInput(type, key)};`);
    }
    return result;
  }

  private getTypedInput(type: Type, key: string): string {
    switch (type) {
      case Type.STRING: {
        return `core.getInput('${key}')`;
      }
      case Type.NUMBER: {
        return `Number(core.getInput('${key}'))`;
      }
      case Type.STRING_LIST: {
        this.includeListFunction = true;
        return `${LIST_FUNCTION}(core.getInput('${key}'))`;
      }
      case Type.NUMBER_LIST: {
        this.includeNumberFunction = true;
        return `${NUMBER_FUNCTION}(${LIST_FUNCTION}(core.getInput('${key}')))`;
      }
      case Type.JSON: {
        return `JSON.parse(core.getInput('${key}'))`;
      }
    }
  }

  private renderOptions(ind: string): string[] {
    const result = [];
    for (const prop of Object.values(this.properties)) {
      for (const doc of docString(prop.description, prop.default)) {
        result.push(doc);
      }
      result.push(`${ind}readonly ${prop.name}${prop.required ? '' : '?'}: ${prop.type};`);
    }
    return result;

    function docString(desc: string, def?: string): string[] {
      const doc = [
        `${ind}/**`,
        `${ind} * ${desc}`,
      ];

      if (def) {
        doc.push(`${ind} *`);
        doc.push(`${ind} * @default - ${def}`);
      }

      doc.push(`${ind} */`);
      return doc;
    }
  }
}

function sanitize(name: string): string {
  const words = name.split('-');
  words[0].charAt(0).toLocaleLowerCase();
  for (let i = 1; i < words.length; i++) {
    words[i].charAt(0).toLocaleUpperCase();
  }
  return words.join('');
}

const LIST_FUNCTION = 'renderListInput';
function renderListFunction(include: boolean): string[] {
  if (!include) { return []; }
  return [
    '',
    '/**',
    ' * Renders a TypeScript list based on what we expect the list to look like in yaml.',
    ' * We expect to see something like "[item1,item2]". GitHub will return \'\' if the input',
    ' * is not defined, which will get translated by this function to [].',
    ' */',
    `function ${LIST_FUNCTION}(rawInput: string): string[] {`,
    '  return (rawInput === \'\' || rawInput === \'[]\') ? [] : rawInput.replace(/\\[|\\]/gi, \'\').split(\',\');',
    '}',
  ];
}

const NUMBER_FUNCTION = 'toNumber';
function renderNumberFunction(include: boolean): string[] {
  if (!include) { return []; }
  return [
    '',
    `function ${NUMBER_FUNCTION}(list: string[]): number[] {`,
    '  return list.map((i) => Number(i));',
    '}',
  ];
}

function indent(len: number): string {
  return '  '.repeat(len);
}
