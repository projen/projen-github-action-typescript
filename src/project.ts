import * as path from 'path';
import { SourceCode, typescript, YamlFile } from 'projen';
import { GitHubActionMetadata } from './github-action-metadata';
import { RunsUsing } from './model/actions-metadata-model';
import { SampleCode } from './sample-code';
import { GitHubActionSourceCode } from './source-code';

/**
 * Properties for creating a GitHubActionTypeScriptProject.
 */
export interface GitHubActionTypeScriptOptions extends typescript.TypeScriptProjectOptions {
  /**
   * Options for the GitHub Action metadata stored in `action.yml`.
   *
   * @default - an action named after the project `name` that runs from `dist/index.js`.
   */
  readonly actionMetadata?: GitHubActionMetadata;

  /**
   * Let Projen manage index.ts and action-options.ts source code.
   *
   * @default true
   */
  readonly manageEntryFiles?: boolean;
}

/**
 * Create a GitHub Action with TypeScript.
 *
 * @pjid github-action-ts
 */
export class GitHubActionTypeScriptProject extends typescript.TypeScriptProject {
  public actionMetadata: GitHubActionMetadata;

  constructor(options: GitHubActionTypeScriptOptions) {
    super({
      ...options,
      sampleCode: false,
    });

    // standard GitHub action packages
    this.addDeps('@actions/core', '@actions/github');

    // package as a single runnable .js file in /dist
    this.addDevDeps('@vercel/ncc');
    this.packageTask.reset('ncc build --source-map --license licenses.txt');

    this.package.addField('main', 'lib/index.js');
    this.addGitIgnore('!/dist/');
    this.annotateGenerated('/dist/**');

    // Create metadata for projen managed `action.yml` file.
    const defaultMetadataOptions: GitHubActionMetadata = {
      name: this.name,
      description: `A GitHub Action for ${this.name}`,
      runs: {
        using: RunsUsing.NODE_16,
        main: 'dist/index.js',
      },
    };

    this.actionMetadata = {
      ...defaultMetadataOptions,
      ...options.actionMetadata,
    };

    new YamlFile(this, 'action.yml', {
      obj: this.actionMetadata,
    });

    if (options.sampleCode ?? true) {
      // If you don't let projen manage files for you, we'll still add it as sample code
      new SampleCode(this, !options.manageEntryFiles);
    }

    // Let projen manage index.ts and action-options.ts files for you
    if (options.manageEntryFiles ?? true) {
      const source = new GitHubActionSourceCode(this);
      const indexSrc = new SourceCode(this, path.join(this.srcdir, 'index.ts'));
      const actionOptionsSrc = new SourceCode(this, path.join(this.srcdir, 'action-options.ts'));
      renderSource(indexSrc, source.indexCode);
      renderSource(actionOptionsSrc, source.actionOptionsCode);
    }
  }
}

function renderSource(src: SourceCode, content: string[]) {
  if (src.marker) {
    src.line(`// ${src.marker}`);
  }

  for (const line of content) {
    src.line(line);
  }
}