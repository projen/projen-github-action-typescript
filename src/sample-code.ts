import { Component, SampleDir } from 'projen';
import { GitHubActionTypeScriptProject } from './project';
import { GitHubActionSourceCode } from './source-code';

export class SampleCode extends Component {
  constructor(project: GitHubActionTypeScriptProject, additionalFiles?: boolean) {
    super(project);

    const actionCode = [
      '// import * as github from \'@actions/github\';',
      'import { ActionManagerOptions } from \'./action-options\';',
      '',
      'export class ActionManager {',
      '',
      '  constructor(options: ActionManagerOptions) {',
      '    console.log(options);',
      '  }',
      '',
      '  main() {',
      '    // this is your action\'s entrypoint',
      '  }',
      '}',
    ].join('\n');

    if (additionalFiles) {
      const sourceCode = new GitHubActionSourceCode(project);
      new SampleDir(project, project.srcdir, {
        files: {
          'action.ts': actionCode,
          'index.ts': sourceCode.indexCode.join('\n'),
          'action-options.ts': sourceCode.actionOptionsCode.join('\n'),
        },
      });
    } else {
      new SampleDir(project, project.srcdir, {
        files: {
          'action.ts': actionCode,
        },
      });
    }
  }
}