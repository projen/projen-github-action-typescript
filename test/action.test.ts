import { Testing } from 'projen';
import { Type } from '../src/github-action-metadata';
import { RunsUsing } from '../src/model/actions-metadata-model';
import { GitHubActionTypeScriptOptions, GitHubActionTypeScriptProject } from '../src/project';

describe('GitHubActionTypeScriptProject', () => {
  test('default action metadata', () => {
    const project = new GitHubActionTypeScriptProject({
      name: 'test',
      defaultReleaseBranch: 'main',
    });

    const snapshot = Testing.synth(project);
    const workflow = snapshot['action.yml'];
    expect(workflow).toContain(
      'runs:\n  using: node16\n  main: dist/index.js',
    );
  });

  test('custom action metadata', () => {
    const project = new GitHubActionTypeScriptProject({
      name: 'test',
      defaultReleaseBranch: 'main',
      actionMetadata: {
        runs: {
          using: RunsUsing.NODE_12,
          main: 'dist/index.js',
        },
        inputs: {
          input1: {
            description: 'my first input',
          },
        },
        outputs: {
          output1: {
            description: 'my first output',
          },
        },
      },
    });

    const snapshot = Testing.synth(project);
    const workflow = snapshot['action.yml'];
    expect(workflow).toContain(
      'runs:\n  using: node12\n  main: dist/index.js',
    );
    expect(workflow).toContain(
      'inputs:\n  input1:\n    description: my first input',
    );
  });

  describe('source code', () => {
    test('default source code', () => {
      const project = sampleProject();

      const snapshot = Testing.synth(project);
      const generatedIndex = snapshot['src/index.ts'];
      expect(generatedIndex).toMatchSnapshot();
      const generatedActionOptions = snapshot['src/options.generated.ts'];
      expect(generatedActionOptions).toMatchSnapshot();
    });

    test('refuse projen managed source code', () => {
      const project = sampleProject({
        manageOptionsFile: false,
      });

      // same sample code, but not projen managed
      const snapshot = Testing.synth(project);
      const generatedIndex = snapshot['src/index.ts'];
      expect(generatedIndex).toMatchSnapshot();
      const generatedActionOptions = snapshot['src/options.generated.ts'];
      expect(generatedActionOptions).toMatchSnapshot();
    });

    test('refuse sample code altogether', () => {
      const project = sampleProject({
        sampleCode: false,
        manageOptionsFile: false,
      });

      // no sample code generated
      const snapshot = Testing.synth(project);
      expect(snapshot['src/index.ts']).toBeUndefined();
      expect(snapshot['src/options.generated.ts']).toBeUndefined();
    });
  });
});

function sampleProject(overrides: Partial<GitHubActionTypeScriptOptions> = {}): GitHubActionTypeScriptProject {
  return new GitHubActionTypeScriptProject({
    name: 'test',
    defaultReleaseBranch: 'main',
    actionMetadata: {
      runs: {
        using: RunsUsing.NODE_12,
        main: 'dist/index.js',
      },
      inputs: {
        input1: {
          description: 'my first input',
          required: true,
          type: Type.STRING,
        },
        input2: {
          description: 'my second input',
          default: '0',
          type: Type.NUMBER,
        },
        input3: {
          description: 'my third input',
          default: '[first, second]',
          type: Type.STRING_LIST,
        },
        input4: {
          description: 'my fourth input',
          required: true,
          type: Type.NUMBER_LIST,
        },
        input5: {
          description: 'my fifth input',
          default: '{}',
          type: Type.JSON,
        },
      },
    },
    ...overrides,
  });
}