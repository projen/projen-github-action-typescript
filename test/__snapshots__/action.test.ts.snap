// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubActionTypeScriptProject source code default source code 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as core from '@actions/core';
import { ActionManager } from './action';

async function run() {
  const input1: string = core.getInput('input1');
  const input2: number = Number(core.getInput('input2'));
  const input3: string[] = renderListInput(core.getInput('input3'));
  const input4: number[] = toNumber(renderListInput(core.getInput('input4')));
  const input5: Record<string,string> = JSON.parse(core.getInput('input5'));
  const action = new ActionManager({
    input1,
    input2,
    input3,
    input4,
    input5,
  });

  await action.main();
}

/**
 * Renders a TypeScript list based on what we expect the list to look like in yaml.
 * We expect to see something like \\"[item1,item2]\\". GitHub will return '' if the input
 * is not defined, which will get translated by this function to [].
 */
function renderListInput(rawInput: string): string[] {
  return (rawInput === '' || rawInput === '[]') ? [] : rawInput.replace(/\\\\[|\\\\]/gi, '').split(',');
}

function toNumber(list: string[]): number[] {
  return list.map((i) => Number(i));
}

run().catch(error => {
  core.setFailed(error.message);
});"
`;

exports[`GitHubActionTypeScriptProject source code default source code 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
export interface ActionManagerOptions {
  /**
   * my first input
   */
  readonly input1: string;
  /**
   * my second input
   *
   * @default - 0
   */
  readonly input2?: number;
  /**
   * my third input
   *
   * @default - [first,second]
   */
  readonly input3?: string[];
  /**
   * my fourth input
   */
  readonly input4: number[];
  /**
   * my fifth input
   *
   * @default - {}
   */
  readonly input5?: Record<string,string>;
}"
`;

exports[`GitHubActionTypeScriptProject source code default source code 3`] = `
"// import * as github from '@actions/github';
import { ActionManagerOptions } from './action-options';

export class ActionManager {

  constructor(options: ActionManagerOptions) {
    console.log(options);
  }

  main() {
    // this is your action's entrypoint
  }
}"
`;

exports[`GitHubActionTypeScriptProject source code refuse projen managed source code 1`] = `
"import * as core from '@actions/core';
import { ActionManager } from './action';

async function run() {
  const input1: string = core.getInput('input1');
  const input2: number = Number(core.getInput('input2'));
  const input3: string[] = renderListInput(core.getInput('input3'));
  const input4: number[] = toNumber(renderListInput(core.getInput('input4')));
  const input5: Record<string,string> = JSON.parse(core.getInput('input5'));
  const action = new ActionManager({
    input1,
    input2,
    input3,
    input4,
    input5,
  });

  await action.main();
}

/**
 * Renders a TypeScript list based on what we expect the list to look like in yaml.
 * We expect to see something like \\"[item1,item2]\\". GitHub will return '' if the input
 * is not defined, which will get translated by this function to [].
 */
function renderListInput(rawInput: string): string[] {
  return (rawInput === '' || rawInput === '[]') ? [] : rawInput.replace(/\\\\[|\\\\]/gi, '').split(',');
}

function toNumber(list: string[]): number[] {
  return list.map((i) => Number(i));
}

run().catch(error => {
  core.setFailed(error.message);
});"
`;

exports[`GitHubActionTypeScriptProject source code refuse projen managed source code 2`] = `
"export interface ActionManagerOptions {
  /**
   * my first input
   */
  readonly input1: string;
  /**
   * my second input
   *
   * @default - 0
   */
  readonly input2?: number;
  /**
   * my third input
   *
   * @default - [first,second]
   */
  readonly input3?: string[];
  /**
   * my fourth input
   */
  readonly input4: number[];
  /**
   * my fifth input
   *
   * @default - {}
   */
  readonly input5?: Record<string,string>;
}"
`;

exports[`GitHubActionTypeScriptProject source code refuse projen managed source code 3`] = `
"// import * as github from '@actions/github';
import { ActionManagerOptions } from './action-options';

export class ActionManager {

  constructor(options: ActionManagerOptions) {
    console.log(options);
  }

  main() {
    // this is your action's entrypoint
  }
}"
`;
